### ISSUE TITLE
tensorflow.nn.dynamic_rnn with variable as init_state cannot work with Estimator.train

### ISSUE THREAD

**Issue by iron9light** [tensorflow/tensorflow#19084]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Windows10
- **TensorFlow installed from (source or binary)**:binary
- **TensorFlow version (use command below)**:1.8.0
- **Python version**: 3.6
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem
I wrote my own function to generate the init state variable

```python
def get_initial_cell_state(cell, batch_size, dtype):
  state_size = cell.state_size
  i = 0
  def get_state_shape(s):
    c = _concat(1, s, static=True)
    nonlocal i
    name = "init_state_" + str(i)
    i = i + 1
    size = tf.get_variable(name, shape=c, dtype=dtype, initializer=tf.initializers.zeros)
    size = tf.tile(size, [batch_size] + [1] * (len(c) - 1))
    return size
  return nest.map_structure(get_state_shape, state_size)
```

And use it as below:

```python
  rnn_output, _ = tf.nn.dynamic_rnn(
    cell=cell,
    inputs=inputs,
    initial_state=get_initial_cell_state(cell, batch_size=batch_size, dtype=tf.float32),
    parallel_iterations=128,
    dtype=tf.float32
  )
```

It's compilable and runnable. But got this error:

`WARNING:tensorflow:It seems that global step (tf.train.get_global_step) has not been increased. Current value (could be stable): 0 vs previous value: 0. You could increase the global step by passing tf.train.get_global_step() to Optimizer.apply_gradients or Optimizer.minimize.`

I use Estimator to train. If I don't use this init state variable, everything goes fine.

**Comment by iron9light** [comment]
Same issue when using tensorflow 1.7.0

**Comment by angersson** [comment]
@ebrevdo Can you take a look at this?

**Comment by iron9light** [comment]
@angersson @ebrevdo Any insight? Or Is there any workable sample about making initial state of RNN trainable?
Trainable initial state will improve performance of RNN model, I think it's a very basic and important feature.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @angersson: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @angersson: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @angersson: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @angersson: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ispirmustafa** [comment]
I don't have an idea.
@iron9light could you please provide a simple script which reproduces the issue?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by iron9light** [comment]
@ispirmustafa Tired. I tried to reproduce this issue with a simple project, but everything goes fine. My real project still has this issue with tensorflow 2.10.

Anyway, you can think it's a feature request:
Add an `init_state_trainable: bool` as the param of `nn.dynamic_rnn` or make it a feature of `nn.rnn_cell.RNNCell` class

**Comment by qlzh727** [comment]
It seems to be a bit strange use case for having initial state as variable. Can u describe what's the intention behind that? eg u want to use those variable in a later step/layer?

**Comment by iron9light** [comment]
https://r2rt.com/non-zero-initial-states-for-recurrent-neural-networks.html

**Comment by qlzh727** [comment]
Sorry for the very late reply. If the initial state need to be a variable, I think it should be attached to the cell itself. So the workaround could be:

```
def create_initial_cell_state(cell, inputs, dtype=tf.float32):
  state_size = cell.state_size
  batch = inputs.shape[0].value

  cell._initial_state_weight = cell.add_variable(
      "init_state", shape=[batch, state_size], initializer=tf.initializers.zeros, dtype=dtype)

  return cell._initial_state_weight

tf.nn.dynamic_rnn(
        cell=cell,
        inputs=x,
        dtype=tf.float32,
        initial_state=create_initial_cell_state(cell, x) if self._rnn_init_state_trainable else None
)
```

This should work for the simple cell state, like u give in your sample code for GRU. LSTM might need some updates since it has tuple state.

I ran some local test and it works without the warning message.

**Comment by iron9light** [comment]
Your solution is nearly the same as my sample code (my sample code works fine).
My original issue is that, this way cannot work in my produce code, but works with the sample code.
What I mean ask for a feature, is make it more easier to do this kinda things by making it a parameter of nn.dynamic_rnn function "init_state_trainable: bool"

**Comment by qlzh727** [comment]
Without looking at your other production code, we won't be able to identify the problem. Since you are requesting a feature, change the issue from bug to FR.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @qlzh727, @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @qlzh727: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @qlzh727: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @qlzh727: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @qlzh727: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmothukuru** [comment]
@iron9light,
Sorry for the delayed response. As [tf.compat.v1.nn.dynamic_rnn
](https://www.tensorflow.org/api_docs/python/tf/compat/v1/nn/dynamic_rnn) has been deprecated in favor of [tf.keras.layers.RNN](https://www.tensorflow.org/api_docs/python/tf/keras/layers/RNN)
 and as **`Tensorflow 2.x Versions`** pre-dominantly use [Tensorflow Keras](https://www.tensorflow.org/api_docs/python/tf/keras), and not [Estimators](https://www.tensorflow.org/guide/estimator), can you please let us know if this issue is still relevant? Thanks!
